shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -351,16 +340,27 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
 
   private[rapids] def compressColumnarBatchWithParquet(
      gpuCB: ColumnarBatch): List[ParquetCachedBatch] = {
-    // NOTE: this doesn't take nulls into account so we could be over-estimating the actual size
-    // but that's still better than under-estimating
-    val estimatedRowSize = scala.Range(0, gpuCB.numCols()).map { index =>
-      gpuCB.column(index).dataType().defaultSize
+    val estimatedRowSize = scala.Range(0, gpuCB.numCols()).map { idx =>
+      gpuCB.column(idx).asInstanceOf[GpuColumnVector].getBase.getDeviceMemorySize / gpuCB.numRows()
     }.sum
     val rowsInBatch = (BYTES_ALLOWED_PER_BATCH / estimatedRowSize).toInt
     val splitIndices = scala.Range(rowsInBatch, gpuCB.numRows(), rowsInBatch)
-    val cudfTables = if (splitIndices.nonEmpty) {
-      val splutVectors = scala.Range(0, gpuCB.numCols()).map { index =>
-        gpuCB.column(index).asInstanceOf[GpuColumnVector].getBase.split(splitIndices: _*)
+    val buffers = new ListBuffer[ParquetCachedBatch]
+    if (splitIndices.nonEmpty) {
+      val splitVectors = new ListBuffer[Array[ColumnVector]]
+      for (index <- 0 until gpuCB.numCols()) {
+        try {
+          splitVectors +=
+            gpuCB.column(index).asInstanceOf[GpuColumnVector].getBase.split(splitIndices: _*)
+        } catch {
+          case _ => splitVectors.foreach { array =>
+            array.foreach { vector =>
+              if (vector != null) {
+                vector.close()
+              }
```

Comment:
I am not sure this is correct. I almost want to rethrow the exception here so the caller knows things went wrong. Or maybe throw another exception?

Reply:
```
        try {
          splitVectors +=
            gpuCB.column(index).asInstanceOf[GpuColumnVector].getBase.split(splitIndices: _*)
        } catch {
          case e => splitVectors.foreach { array =>
            array.foreach { vector =>
              if (vector != null) {
                vector.close()
              }
            }
          }
            // the caller will want to know there was an exception but won't be able to close the
            // splitVectors
            throw e
        }
```

Reply:
I confused myself with the log throw anti-pattern. I have made the above mentioned fix

Reply:
The code how you have it looks correct to me.  You should re-throw the exception in the case of an error. The indentation appears to be off but that is minor.

As an alternative because when you create a table it will increment the reference count for the columns passed in, you could just use a try/finally block for splitVectors.

```
val splitVectors =  new ListBuffer[Array[ColumnVector]]
try {
  for (index <- 0 until gpuCB.numCols()) {
    splitVectors +=
        gpuCB.column(index).asInstanceOf[GpuColumnVector].getBase.split(splitIndices: _*)
  }

  def makeTableForIndex(i: Int): Table = {
    val columns = splitVectors.indices.map(j => splitVectors(j)(i))
    new Table(columns :_*)
  }

  for (i <- splitVectors(0).indices) {
    withResource(makeTableForIndex(i)) { table =>
      buffers += ParquetCachedBatch(writeTableToCachedBatch(table))
    }
  }
} finally {
  // Not 100% sure the safeClose intrinsic works here
  splitVectors.foreach(array => array.safeClose())
}
```